File: C++/nth_node_from_end_of_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = head;
        int N = 0;

        while(temp != nullptr){
            N++;
            temp = temp->next;
        }

        if(n == N){return head->next;}
        temp = head;
        int i = 0;

        while(true){
            if (i == N - n - 1){
                temp->next = temp->next->next;
                return head;
            }
            i++;
            temp = temp->next;
        }        

        return nullptr;
    }
};

File: python/nth_node_from_end_of_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        N = 0
        temp = head
        while temp:
            N += 1
            temp = temp.next
        
        if N == n:
            return head.next
        
        temp = head
        i = 0
        while temp:
            if i == N - n - 1:
                temp.next = temp.next.next
                return head
            temp = temp.next
            i += 1
